File: packages/server/src/bootstrap/registry.ts
/**
 * @fileoverview Bootstrap commands registry
 *
 * Provides a function to get all bootstrap commands for an AFD server.
 */

import { z } from 'zod';
import { defineCommand, type ZodCommandDefinition } from '../schema.js';
import type { ContextConfig } from '../server-types.js';
import {
	type ContextState,
	createAfdContextEnterCommand,
	createAfdContextExitCommand,
	createAfdContextListCommand,
} from './afd-context.js';
import { createAfdDocsCommand } from './afd-docs.js';
import { createAfdHelpCommand } from './afd-help.js';
import { createAfdSchemaCommand } from './afd-schema.js';
import type { DescribedCommand, GetDescribedCommands } from './described-command.js';

/** Names of the discovery bootstrap tools registered by `createMcpServer({ bootstrap: true })`. */
export const BOOTSTRAP_COMMAND_NAMES: readonly string[] = ['afd-help', 'afd-docs', 'afd-schema'];

/** Names of the context tools registered when `createMcpServer({ contexts })` is set. */
export const CONTEXT_COMMAND_NAMES: readonly string[] = [
	'afd-context-list',
	'afd-context-enter',
	'afd-context-exit',
];

/**
 * Get all bootstrap commands for an AFD server: afd-help, afd-docs and afd-schema,
 * plus the afd-context-* commands when `contexts` and `contextState` are given.
 *
 * Every returned command is a `ZodCommandDefinition` with `expose: { mcp: true }`.
 * The discovery tools describe only the MCP-exposed commands that `getCommands`
 * returns. With `createMcpServer`, prefer the `bootstrap: true` option, which also
 * lists the built-in and context commands.
 *
 * @param getCommands - Function to get the commands to describe
 * @param options - Optional configuration
 * @returns Array of bootstrap command definitions
 */
export function getBootstrapCommands(
	getCommands: GetDescribedCommands,
	options?: {
		getJsonSchema?: (cmd: DescribedCommand) => Record<string, unknown>;
		contexts?: ContextConfig[];
		contextState?: ContextState;
	}
): ZodCommandDefinition[] {
	const cmds: ZodCommandDefinition[] = [
		createAfdHelpCommand(getCommands),
		createAfdDocsCommand(getCommands),
		createAfdSchemaCommand(getCommands, options?.getJsonSchema),
	];

	// Add context commands when contexts are configured
	if (options?.contexts?.length && options.contextState) {
		const getContexts = () => options.contexts ?? [];
		const contextState = options.contextState;
		cmds.push(
			defineCommand({
				...createAfdContextListCommand(getContexts, contextState),
				input: z.object({}),
				expose: { mcp: true },
			}),
			defineCommand({
				...createAfdContextEnterCommand(getContexts, contextState),
				input: z.object({ context: z.string().min(1).describe('Context name to enter') }),
				expose: { mcp: true },
			}),
			defineCommand({
				...createAfdContextExitCommand(contextState),
				input: z.object({}),
				expose: { mcp: true },
			})
		);
	}

	return cmds;
}

File: python/src/afd/server/bootstrap/registry.py
"""Bootstrap commands registry.

Provides a function to get all bootstrap commands for an AFD server.
"""

from typing import Any, Callable, Dict, List, Optional

from afd.core.commands import CommandDefinition
from afd.server.bootstrap.afd_help import create_afd_help_command
from afd.server.bootstrap.afd_docs import create_afd_docs_command
from afd.server.bootstrap.afd_schema import create_afd_schema_command


def get_bootstrap_commands(
    get_commands: Callable[[], List[CommandDefinition]],
    options: Optional[Dict[str, Any]] = None,
) -> List[CommandDefinition]:
    """Get all bootstrap commands for an AFD server.

    Bootstrap commands provide built-in functionality for command discovery
    and documentation:
    - afd-help: List all available commands with filtering and grouping
    - afd-docs: Generate markdown documentation for commands
    - afd-schema: Export JSON schemas for all commands

    Args:
        get_commands: Function to get all user-defined commands.
        options: Optional configuration:
            - get_json_schema: Function to get JSON schema for a command.

    Returns:
        List of bootstrap command definitions.

    Example:
        >>> from afd.server.bootstrap import get_bootstrap_commands
        >>> from afd.core.commands import create_command_registry
        >>>
        >>> registry = create_command_registry()
        >>> # ... register user commands ...
        >>>
        >>> bootstrap = get_bootstrap_commands(registry.list)
        >>> for cmd in bootstrap:
        ...     registry.register(cmd)
    """
    options = options or {}
    get_json_schema = options.get("get_json_schema")

    return [
        create_afd_help_command(get_commands),
        create_afd_docs_command(get_commands),
        create_afd_schema_command(get_commands, get_json_schema),
    ]
